Libplanet/Action/AccountStateDeltaImpl.cs

```diff
@@ -1,53 +1,198 @@
+                ? value
+                : _accountStateGetter(address);
+
+        /// <inheritdoc/>
+        [Pure]
+        IAccountStateDelta IAccountStateDelta.SetState(Address address, IValue state) =>
+            UpdateStates(_updatedStates.SetItem(address, state));
+
+        /// <inheritdoc/>
+        [Pure]
+        public BigInteger GetBalance(Address address, Currency currency) =>
+            _updatedFungibleAssets.TryGetValue((address, currency), out BigInteger balance)
+                ? balance
+                : _accountBalanceGetter(address, currency);
+
+        /// <inheritdoc/>
+        [Pure]
+        public IAccountStateDelta MintAsset(Address recipient, Currency currency, BigInteger amount)
         {
-            try
+            if (amount <= 0)
             {
-                return _updatedStates[address];
+                throw new ArgumentOutOfRangeException(
+                    nameof(amount),
+                    "The amount to mint has to be greater than 0."
+                );
             }
-            catch (KeyNotFoundException)
+            else if (!currency.AllowsToMint(_signer))
             {
-                return _accountStateGetter(address);
+                throw new CurrencyPermissionException(
+                    _signer,
+                    currency,
+                    $"The account {_signer} has no permission to mint the currency {currency}."
+                );
             }
+
+            BigInteger balance = GetBalance(recipient, currency);
+            return UpdateFungibleAssets(
+                _updatedFungibleAssets.SetItem((recipient, currency), balance + amount)
+            );
         }
 
         /// <inheritdoc/>
-        IAccountStateDelta IAccountStateDelta.SetState(Address address, IValue state)
+        [Pure]
+        public IAccountStateDelta TransferAsset(
+            Address sender,
+            Address recipient,
+            Currency currency,
+            BigInteger amount,
+            bool allowNegativeBalance = false
+        )
         {
-            return new AccountStateDeltaImpl(_accountStateGetter)
+            if (amount <= 0)
             {
-                _updatedStates = _updatedStates.SetItem(address, state),
-            };
+                throw new ArgumentOutOfRangeException(
+                    nameof(amount),
+                    "The amount to transfer has to be greater than 0."
+                );
+            }
+
+            BigInteger senderBalance = GetBalance(sender, currency),
+                       recipientBalance = GetBalance(recipient, currency);
+
+            if (!allowNegativeBalance && senderBalance < amount)
+            {
+                var msg = $"The account {sender}'s balance of {currency} is insufficient to " +
+                          $"transfer: {senderBalance} {currency} < {amount} {currency}.";
+                throw new InsufficientBalanceException(sender, currency, 0, msg);
```

Comment:
Shouldn't this be the sender's balance?
```suggestion
                throw new InsufficientBalanceException(sender, currency, senderBalance, msg);
```